src/sleep.rs

```
@@ -71,7 +71,7 @@ impl Future for Sleep {
                 return Poll::Ready(());
             }
 
-            let duration_ms: i32 = (this.deadline - now).as_millis().try_into().unwrap();
+            let duration_ms: i32 = (this.deadline - now).as_millis().try_into().unwrap_or(i32::MAX);
```

IMO, it should not silently replace the value. If the value is larger than `i32::MAX`, then:
 - either `this.deadline < now` (and the subtraction overflows), which is an error
 - or `this.deadline` is more than 24 days after `now`, and it is better to panic than replace the value silently

---

The problem is that it's helpful to be able to sleep for `Duration::MAX` as a sentinel value or a value to be replaced.  It's a bit scary to have a non-deterministic panic, and it also doesn't match Tokio's behaviour (which doesn't crash when sleeping for `Duration::MAX`).

---

Yes, they added some `Instant::far_future()`: https://github.com/tokio-rs/tokio/commit/d2ad7afd21e4faef05ccccb0288bc814a954b990

Will think about it.

(Just for info, to wait indefinitely, there's `future::pending()`.)

---

Right, but that requires a timeout that might be indefinite to construct a type of futures that might be `Sleep` or might be `Pending`.  It's workable, but much less ergonomic than supporting (i.e. without crashing) an upper bound on `Duration` (which exists already in the `std::time` API).

---

> Will think about it.

OK, I'm convinced. I will merge this commit: 633ff1e8738bbc44b04d3ce44a1d86da519d6f64.

---

That looks great, thank you!